### PR TITLE
ARROW-5413: [C++] Skip UTF8 BOM in CSV files

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

File: cpp/src/arrow/util/utf8.cc [hunk 1]
```diff
@@ -79,5 +79,31 @@ void InitializeUTF8() {
   std::call_once(utf8_initialized, internal::InitializeLargeTable);
 }
 
+static const uint8_t kBOM[] = "\xef\xbb\xbf";
```
Comment (fsaintjacques) [inline cpp/src/arrow/util/utf8.cc]: string to `uint8_t[]` seems odd (and the -1 in the loop), why not use `static const uint8_t kBOM[] = [0xEF, 0xBB, 0xBF]`?

File: cpp/src/arrow/util/utf8.cc [hunk 2]
```diff
@@ -79,5 +79,31 @@ void InitializeUTF8() {
...
+Status SkipUTF8BOM(const uint8_t* data, int64_t size, const uint8_t** out) {
+  int64_t i;
+  for (i = 0; i < static_cast<int64_t>(sizeof(kBOM) - 1); ++i) {
+    if (size == 0) {
+      if (i == 0) {
+        // Empty string
+        *out = data;
+        return Status::OK();
+      } else {
+        return Status::Invalid("UTF8 string too short (truncated byte order mark?)");
```
Comment (fsaintjacques) [inline cpp/src/arrow/util/utf8.cc]: Should this be an error or should it just avoid skipping it? I have no good enough knowledge of UTF to make such decision.

**Review by fsaintjacques — COMMENTED** [review]
LGTM, minor comments.

File: cpp/src/arrow/util/utf8.cc (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/util/utf8.cc]: Ah, right, I hadn't thought about that.

File: cpp/src/arrow/util/utf8.cc (hunk 2 above)
Comment (pitrou) [reply cpp/src/arrow/util/utf8.cc]: If the streams ends there, it's a truncated UTF8 character, so it's invalid.

**Review by kou — APPROVED** [review]
+1

**Pull request closed without merging**